packages/erk-shared/src/erk_shared/github/retry.py

```diff
@@ -1,71 +1,139 @@
-"""Retry utility for GitHub API calls with exponential backoff."""
+"""Retry utilities with configurable delays and return-value control flow."""
+
+from __future__ import annotations
 
 import sys
 from collections.abc import Callable
+from dataclasses import dataclass
 from typing import TypeVar
 
 from erk_shared.gateway.time.abc import Time
 
-# Retry delays following exponential backoff pattern: [0.5, 1.0] (~1.5s max, 2 retries)
+# Default retry delays for transient error retry (exponential backoff)
 RETRY_DELAYS = [0.5, 1.0]
 
 T = TypeVar("T")
 
 
-def with_github_retry(
+@dataclass(frozen=True)
+class RetryRequested:
+    """Callback returns this to signal that retry is needed.
+
+    Used by callbacks to indicate a transient error or that the desired
+    state hasn't been reached yet (eventual consistency polling).
+
+    Attributes:
+        reason: Human-readable description of why retry is needed
+    """
+
+    reason: str
+
+
+@dataclass(frozen=True)
+class RetriesExhausted:
+    """Returned by with_retries when all retry attempts failed.
+
+    Distinct from RetryRequested to clarify semantics:
+    - RetryRequested: callback signals "try again"
+    - RetriesExhausted: with_retries signals "gave up after N attempts"
+
+    Attributes:
+        reason: The reason from the last RetryRequested
+    """
+
+    reason: str
+
+
+def with_retries(
```

Comment:
we should have unit tests for this in isolation for a bunch of different scenarios (where the callable passed in is a toy)

Reply:
Resolved via /erk:pr-address at 2025-12-31 01:13 PM PST: Unit tests already exist in packages/erk-shared/tests/unit/github/test_retry.py with 17 test cases covering various scenarios using toy callables - success on 1st/2nd/3rd attempt, all attempts fail, exception bubbles, custom delays, empty delays, return type preservation, polling patterns, transient errors, and sentinel dataclass behavior.

_Addressed via `/erk:pr-address` at 2025-12-31 13:13 PST_